Source/EasyNetQ/Internals/TaskHelpers.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System;
+using System.Diagnostics;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace EasyNetQ.Internals
+{
+    public static class TaskHelpers
+    {
+        /// <summary>
+        ///     We want to prevent callers hijacking the reader thread; this is a bit nasty, but works;
+        ///     see http://stackoverflow.com/a/22588431/23354 for more information; a huge
+        ///     thanks to Eli Arbel for spotting this (even though it is pure evil; it is *my kind of evil*)
+        /// </summary>
+        private static readonly Func<Task, bool> IsSyncSafe;
+
+        static TaskHelpers()
+        {
+            try
+            {
+                var taskType = typeof (Task);
+                var continuationField = taskType.GetField("m_continuationObject", BindingFlags.Instance | BindingFlags.NonPublic);
```

Comment:
Yes we know is weird, bad and could fail in next framework version, but in first instance I think it's good to go, so to have a nice and good sync/async api that doesn't starve the threadpool :smile: